// wagmi-core/connectors/walletLink/dist/wagmi-core-connectors-walletLink.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var providers = require('@ethersproject/providers');
var walletlink = require('walletlink');
var base = require('../../../dist/base-0267900f.cjs.dev.js');
var utils = require('ethers/lib/utils');
require('eventemitter3');

class WalletLinkConnector extends base.Connector {
  id = 'walletLink';
  name = 'Coinbase Wallet';
  ready = typeof window !== 'undefined' && !window.ethereum?.isCoinbaseWallet;
  #client;
  #provider;

  constructor(config) {
    super(config);
  }

  async connect() {
    try {
      const provider = this.getProvider();
      provider.on('accountsChanged', this.onAccountsChanged);
      provider.on('chainChanged', this.onChainChanged);
      provider.on('disconnect', this.onDisconnect);
      const accounts = await provider.enable();
      const account = utils.getAddress(accounts[0]);
      const id = await this.getChainId();
      const unsupported = this.isChainUnsupported(id);
      return {
        account,
        chain: {
          id,
          unsupported
        },
        provider: new providers.Web3Provider(provider)
      };
    } catch (error) {
      if (/user closed modal/i.test(error.message)) throw new base.UserRejectedRequestError();
      throw error;
    }
  }

  async disconnect() {
    if (!this.#provider) return;
    const provider = this.getProvider();
    provider.removeListener('accountsChanged', this.onAccountsChanged);
    provider.removeListener('chainChanged', this.onChainChanged);
    provider.removeListener('disconnect', this.onDisconnect);
    provider.disconnect();
    provider.close();

    if (typeof localStorage !== 'undefined') {
      let n = localStorage.length;

      while (n--) {
        const key = localStorage.key(n);
        if (!key) continue;
        if (!/-walletlink/.test(key)) continue;
        localStorage.removeItem(key);
      }
    }
  }

  async getAccount() {
    const provider = this.getProvider();
    const accounts = await provider.request({
      method: 'eth_accounts'
    }); // return checksum address

    return utils.getAddress(accounts[0]);
  }

  async getChainId() {
    const provider = this.getProvider();
    const chainId = base.normalizeChainId(provider.chainId);
    return chainId;
  }

  getProvider() {
    if (!this.#provider) {
      this.#client = new walletlink.WalletLink(this.options);
      this.#provider = this.#client.makeWeb3Provider(this.options.jsonRpcUrl);
    }

    return this.#provider;
  }

  async getSigner() {
    const provider = this.getProvider();
    const account = await this.getAccount();
    return new providers.Web3Provider(provider).getSigner(account);
  }

  async isAuthorized() {
    try {
      const account = await this.getAccount();
      return !!account;
    } catch {
      return false;
    }
  }

  async switchChain(chainId) {
    const provider = this.getProvider();
    const id = utils.hexValue(chainId);

    try {
      await provider.request({
        method: 'wallet_switchEthereumChain',
        params: [{
          chainId: id
        }]
      });
      const chains = [...this.chains, ...base.allChains];
      return chains.find(x => x.id === chainId);
    } catch (error) {
      if (/user rejected signature request/i.test(error.message)) throw new base.UserRejectedRequestError();else throw new base.SwitchChainError();
    }
  }

  onAccountsChanged = accounts => {
    if (accounts.length === 0) this.emit('disconnect');else this.emit('change', {
      account: utils.getAddress(accounts[0])
    });
  };
  onChainChanged = chainId => {
    const id = base.normalizeChainId(chainId);
    const unsupported = this.isChainUnsupported(id);
    this.emit('change', {
      chain: {
        id,
        unsupported
      }
    });
  };
  onDisconnect = () => {
    this.emit('disconnect');
  };
}

exports.WalletLinkConnector = WalletLinkConnector;
