// wagmi-core/connectors/walletLink/dist/wagmi-core-connectors-walletLink.esm.js
import { Web3Provider } from '@ethersproject/providers';
import { WalletLink } from 'walletlink';
import { C as Connector, U as UserRejectedRequestError, n as normalizeChainId, b as allChains, S as SwitchChainError } from '../../../dist/base-5c99d89b.esm.js';
import { getAddress, hexValue } from 'ethers/lib/utils';
import 'eventemitter3';

class WalletLinkConnector extends Connector {
  id = 'walletLink';
  name = 'Coinbase Wallet';
  ready = typeof window !== 'undefined' && !window.ethereum?.isCoinbaseWallet;
  #client;
  #provider;

  constructor(config) {
    super(config);
  }

  async connect() {
    try {
      const provider = this.getProvider();
      provider.on('accountsChanged', this.onAccountsChanged);
      provider.on('chainChanged', this.onChainChanged);
      provider.on('disconnect', this.onDisconnect);
      const accounts = await provider.enable();
      const account = getAddress(accounts[0]);
      const id = await this.getChainId();
      const unsupported = this.isChainUnsupported(id);
      return {
        account,
        chain: {
          id,
          unsupported
        },
        provider: new Web3Provider(provider)
      };
    } catch (error) {
      if (/user closed modal/i.test(error.message)) throw new UserRejectedRequestError();
      throw error;
    }
  }

  async disconnect() {
    if (!this.#provider) return;
    const provider = this.getProvider();
    provider.removeListener('accountsChanged', this.onAccountsChanged);
    provider.removeListener('chainChanged', this.onChainChanged);
    provider.removeListener('disconnect', this.onDisconnect);
    provider.disconnect();
    provider.close();

    if (typeof localStorage !== 'undefined') {
      let n = localStorage.length;

      while (n--) {
        const key = localStorage.key(n);
        if (!key) continue;
        if (!/-walletlink/.test(key)) continue;
        localStorage.removeItem(key);
      }
    }
  }

  async getAccount() {
    const provider = this.getProvider();
    const accounts = await provider.request({
      method: 'eth_accounts'
    }); // return checksum address

    return getAddress(accounts[0]);
  }

  async getChainId() {
    const provider = this.getProvider();
    const chainId = normalizeChainId(provider.chainId);
    return chainId;
  }

  getProvider() {
    if (!this.#provider) {
      this.#client = new WalletLink(this.options);
      this.#provider = this.#client.makeWeb3Provider(this.options.jsonRpcUrl);
    }

    return this.#provider;
  }

  async getSigner() {
    const provider = this.getProvider();
    const account = await this.getAccount();
    return new Web3Provider(provider).getSigner(account);
  }

  async isAuthorized() {
    try {
      const account = await this.getAccount();
      return !!account;
    } catch {
      return false;
    }
  }

  async switchChain(chainId) {
    const provider = this.getProvider();
    const id = hexValue(chainId);

    try {
      await provider.request({
        method: 'wallet_switchEthereumChain',
        params: [{
          chainId: id
        }]
      });
      const chains = [...this.chains, ...allChains];
      return chains.find(x => x.id === chainId);
    } catch (error) {
      if (/user rejected signature request/i.test(error.message)) throw new UserRejectedRequestError();else throw new SwitchChainError();
    }
  }

  onAccountsChanged = accounts => {
    if (accounts.length === 0) this.emit('disconnect');else this.emit('change', {
      account: getAddress(accounts[0])
    });
  };
  onChainChanged = chainId => {
    const id = normalizeChainId(chainId);
    const unsupported = this.isChainUnsupported(id);
    this.emit('change', {
      chain: {
        id,
        unsupported
      }
    });
  };
  onDisconnect = () => {
    this.emit('disconnect');
  };
}

export { WalletLinkConnector };
